File: src/skilltree/model/TreeHealth.js
// Pure diagnostics over a SkillTree: counts, cross-branch coupling, redundant
// (transitively implied) edges, and a single 0..100 health score.

const REDUNDANCY_NODE_LIMIT = 1500;

export class TreeHealth {
  static assess(tree) {
    const nodeCount = tree.nodes.length;
    const edgeCount = tree.edges.length;
    const crossBranch = tree.edges.filter((edge) => tree.trunk.edgeKind(edge.from, edge.to) === 'cross-branch').length;
    const redundant = nodeCount > REDUNDANCY_NODE_LIMIT ? 0 : TreeHealth.countRedundantEdges(tree);
    const avgInDegree = Math.round((edgeCount / Math.max(1, nodeCount)) * 100) / 100;

    const crossFrac = edgeCount ? crossBranch / edgeCount : 0;
    const redFrac = edgeCount ? redundant / edgeCount : 0;
    const score = Math.round(Math.max(0, Math.min(100, 100 * (1 - 0.6 * crossFrac - 0.4 * redFrac))));

    return { nodeCount, edgeCount, crossBranch, redundant, avgInDegree, score };
  }

  static countRedundantEdges(tree) {
    const descendantsById = new Map();
    for (const id of [...tree.topoOrder()].reverse()) {
      const descendants = new Set();
      for (const child of tree.childrenOf(id)) {
        descendants.add(child.id);
        for (const descendantId of descendantsById.get(child.id)) descendants.add(descendantId);
      }
      descendantsById.set(id, descendants);
    }

    // an edge is redundant when its target is already reachable through a sibling child
    let redundant = 0;
    for (const edge of tree.edges) {
      const implied = tree.childrenOf(edge.from).some((child) => child.id !== edge.to && descendantsById.get(child.id).has(edge.to));
      if (implied) redundant += 1;
    }
    return redundant;
  }
}

File: domain/TreeHealth.cpp
#include "domain/TreeHealth.h"

#include <algorithm>
#include <cmath>
#include <map>
#include <set>

namespace wm {

namespace {
int countRedundantEdges(const SkillTree& tree) {
  std::map<NodeId, std::set<NodeId>> descendantsById;
  const std::vector<NodeId>& order = tree.topoOrder();
  for (auto it = order.rbegin(); it != order.rend(); ++it) {
    std::set<NodeId> descendants;
    for (const NodeSpec* child : tree.childrenOf(*it)) {
      descendants.insert(child->id);
      for (const NodeId& deep : descendantsById[child->id]) descendants.insert(deep);
    }
    descendantsById[*it] = std::move(descendants);
  }

  int redundant = 0;
  for (const Edge& edge : tree.edges()) {
    for (const NodeSpec* sibling : tree.childrenOf(edge.from)) {
      if (sibling->id != edge.to && descendantsById[sibling->id].count(edge.to)) {
        ++redundant;
        break;
      }
    }
  }
  return redundant;
}
}

Health TreeHealth::assess(const SkillTree& tree) {
  Health health;
  health.nodeCount = static_cast<int>(tree.nodes().size());
  health.edgeCount = static_cast<int>(tree.edges().size());

  for (const Edge& edge : tree.edges()) {
    if (tree.trunk().edgeKind(edge.from, edge.to) == EdgeKind::cross_branch) ++health.crossBranch;
  }
  health.redundant = health.nodeCount > redundancyNodeLimit ? 0 : countRedundantEdges(tree);
  health.avgInDegree = std::round((static_cast<double>(health.edgeCount) / std::max(1, health.nodeCount)) * 100) / 100;

  double crossFrac = health.edgeCount ? static_cast<double>(health.crossBranch) / health.edgeCount : 0;
  double redFrac = health.edgeCount ? static_cast<double>(health.redundant) / health.edgeCount : 0;
  double raw = 100 * (1 - 0.6 * crossFrac - 0.4 * redFrac);
  health.score = static_cast<int>(std::round(std::max(0.0, std::min(100.0, raw))));

  return health;
}

}
